update: custom scene management and InstantiateAndSpawn [MTT-7399] (#1142)
Unity-Technologies/com.unity.multiplayer.docs: docs/basics/scenemanagement/custom-management.md

:::

## Building Your Own Scene Management Solution

> Netcode for GameObjects has primarily focused on providing a scene management solution that should meet most projects' needs. However, there might be a special case scenario where you require building your own scene management solution. The first step is to disable the "Enable Scene Management" property in your `NetworkManager`'s properties. All of your scene loading and unloading has to be handled via the `UnityEngine.SceneManagement.SceneManager` class.  

Reviewer: ```suggestion
Netcode for GameObjects is designed to provide a scene management solution that should meet most projects' needs. However, some projects might still need to build their own scene management solution. The first step is to disable **Enable Scene Management** in your `NetworkManager`'s properties. All of your scene loading and unloading must be handled via the `UnityEngine.SceneManagement.SceneManager` class.  
```

### Integrated NetworkSceneManager Comparison (Enabled vs Disabled)
Network Scene Management      | Enabled  | Disabled
------------------------------|----------|----------

> Scene Event Synchronizing     | Yes      | No
> In-Scene Network Prefab Synch | Yes      | Yes
> In-Scene NetworkObject Synch  | Yes      | No
> Auto scene migration Synch    | Yes      | No

Reviewer: ```suggestion
Scene Event Synchronizing     | Yes      | No
In-Scene Network Prefab Synchronizing | Yes      | Yes
In-Scene NetworkObject Synchronizing  | Yes      | No
Auto Scene Migration Synchonizing    | Yes      | No
```

#### Scene Event Synchronizing

> When integrated scene management is enabled, scene related events are automatically synchronized for you. When integrated scene management is disabled, you need to handle the synchronization of scenes yourself.

Reviewer: ```suggestion
When integrated scene management is enabled, scene-related events are automatically synchronized for you. When integrated scene management is disabled, you need to handle the synchronization of scenes yourself.
```

#### In-Scene Placed Network Prefabs and NetworkObjects

> Both when enabled and disabled, a joining client is synchronized with any in-scene placed Network Prefab instances. When integrated scene management is disabled, only the scenes loaded prior to a client joining will have their in-scene placed Network Prefab instance synchronized. If you load any other scenes after a client is connected, they will not be synchronized (you would want to only dynamically spawn NetworkObjects at that point).
> 
> :::warning Any in-scene defined NetworkObject (i.e. not a network prefab instance) will not get synchronized with a newly joined client when scene management is disabled.

Reviewer: ```suggestion
Clients are always synchronized with any in-scene placed Network Prefab instances when they join a session, regardless of whether integrated scene management is enabled. When scene management is disabled, only scenes loaded prior to a client joining are synchronized — any scenes loaded *after* a client is connected won't be synchronized (you would want to only dynamically spawn NetworkObjects at that point).

:::warning Any in-scene defined NetworkObject (i.e. not a network prefab instance) will not get synchronized with a newly-joined client when scene management is disabled.
```

:::

> When the integrated `NetworkSceneManager` is enabled both in-scene placed network prefab instances and in-scene defined `NetworkObject`s get synchronized and when new scenes are loaded they get synchronized.

Reviewer: ```suggestion
When the integrated `NetworkSceneManager` is enabled, both in-scene placed network prefab instances and in-scene defined `NetworkObject`s get synchronized, and when new scenes are loaded they get synchronized also.
```

> #### Auto (NetworkObject) Scene migration Synchronization
> When integrated scene management is enabled, you can have the server migrate an already spawned `NetworkObject` from one scene to another and it will automatically be synchronized with connected and late joining clients.
> When integrated scene management is disabled, this is not handled and you will need to use either NetworkVariables, Rpcs, or custom messages to handle synchronizing clients with any migration of `NetworkObjects`s from one scene to another.

Reviewer: ```suggestion
#### Auto (NetworkObject) Scene Migration Synchronization
When integrated scene management is enabled, you can have the server migrate an already spawned `NetworkObject` from one scene to another and it will automatically be synchronized with connected and late-joining clients.

When integrated scene management is disabled, this is not handled and you need to use either NetworkVariables, Rpcs, or custom messages to handle synchronizing clients with any migration of `NetworkObjects`s from one scene to another.
```

### Registering In-Scene Placed Network Prefab Instances

> While integrated scene management solution handles the synchronization of in-scene placed `NetworkObject`s, custom scene management treats everything like a dynamically spawned `NetworkObject`.  As such, you **can only use network prefabs defined in your NetworkPrefabList assigned to your NetworkManager and any in-scene defined `NetworkObject`s will not synchronize with clients**.
> 
> Once you have registered your in-scene placed Network Prefabs within your `NetworkPrefabList` and assigned that to your `NetworkManager`, you can then start a server/host and have a client connect and synchronize with the in-scene placed Network Prefab instances **(as long as both client and server have pre-loaded the scene or scenes required)**.  

Reviewer: ```suggestion
While integrated scene management handles the synchronization of in-scene placed `NetworkObject`s, custom scene management treats everything like a dynamically spawned `NetworkObject`.  As such, you can only use network prefabs defined in your NetworkPrefabList assigned to your NetworkManager, and any in-scene defined `NetworkObject`s will not synchronize with clients.

Once you've registered your in-scene placed Network Prefabs with your `NetworkPrefabList` and assigned that to your `NetworkManager`, you can then start a server or host and have a client connect and synchronize with the in-scene placed Network Prefab instances (as long as both client and server have pre-loaded the scene or scenes required).  
```

:::important

> When a client first connects, it will delete any in-scene placed `NetworkObjects` in any of the scenes it has currently loaded.  When using a custom scene management solution, In-scene placed `NetworkObject`s are actually dynamically spawned. This means any changes you make to your in-scene placed Network Prefabs will **not** be synchronized with clients automatically.

Reviewer: ```suggestion
When a client first connects, it deletes any in-scene placed `NetworkObjects` in any of the scenes it has currently loaded.  When using a custom scene management solution, in-scene placed `NetworkObject`s are actually dynamically spawned. This means any changes you make to your in-scene placed Network Prefabs will *not* be synchronized with clients automatically.
```

:::

### Synchronizing In-Scene Placed Network Prefab Instances

> If you would like any changes made to an in-scene placed network prefab instance, then you will need to handle the serialization of these settings yourself. You can do this by overriding `NetworkBehaviour.OnSynchronize` and serializing any property updates you want to have synchronized with clients when they join. [Read More About OnSynchronize Here](../../basics/networkbehaviour.md#pre-spawn-synchronization)

Reviewer: ```suggestion
If you want to change an in-scene placed network prefab instance, you need to handle the serialization of these settings yourself. You can do this by overriding `NetworkBehaviour.OnSynchronize` and serializing any property updates you want to have synchronized with clients when they join. [Read More About OnSynchronize Here](../../basics/networkbehaviour.md#pre-spawn-synchronization).
```

### Scene synchronization
 While you might have an initial set of scenes loaded, you are bound to want to load more scenes as your netcode enabled game session progresses.  Once a client has fully connected (you can use `NetworkManager.OnClientConnected` for this), you will want to send the client a list of additional scenes to be loaded.
  - You will want to come up with your own "client state" as it progresses through this synchronization process to determine when a client has loaded all scenes.  

>     - As an example:  A client might be connected and has synchronized with the default scenes required to connect, but then you have one or more additional scenes you might have loaded (additively) that the client needs to load and synchronize (spawn) any in-scene placed Network Prefab instances

Reviewer: ```suggestion
    - For example:  a client might be connected and has synchronized with the default scenes required to connect, but then you have one or more additional scenes you might have loaded (additively) that the client needs to load and synchronize (spawn) any in-scene placed Network Prefab instances.
```

>   - Remember, once a client has connected, if you load additional scenes in-scene placed network prefab instances will not be synchronized for you. It is recommended to dynamically spawn network prefab instances after the initial client synchronization (if you plan on loading more scenes).

Reviewer: ```suggestion
  - Remember, once a client has connected, if you load additional scenes then any in-scene placed network prefab instances will not be synchronized for you. It's recommended to dynamically spawn network prefab instances after the initial client synchronization if you plan on loading more scenes.
```
